File: src/sampling.py
import torch


def sample_collocation(N, pde, device):
    x = (torch.rand(N, 1, device=device) * (pde.x_max - pde.x_min) + pde.x_min).requires_grad_(True)
    t = (torch.rand(N, 1, device=device) * pde.t_max).requires_grad_(True)
    return x, t


def sample_ic(N, pde, device):
    x = torch.rand(N, 1, device=device) * (pde.x_max - pde.x_min) + pde.x_min
    t = torch.zeros(N, 1, device=device)
    u = pde.initial_condition(x)
    return x, t, u


def sample_bc(N, pde, device):
    t = torch.rand(N, 1, device=device) * pde.t_max
    return t

File: cpp/src/sampling.cpp
#include "sampling.hpp"

std::pair<torch::Tensor, torch::Tensor>
sample_collocation(int N, PDEProblem& pde, torch::Device dev) {
    auto opts = torch::TensorOptions().dtype(torch::kFloat32).device(dev);
    auto x = (torch::rand({N, 1}, opts) * (pde.x_max - pde.x_min) + pde.x_min);
    auto t = (torch::rand({N, 1}, opts) * pde.t_max);
    x.set_requires_grad(true);
    t.set_requires_grad(true);
    return {x, t};
}

std::pair<torch::Tensor, torch::Tensor>
sample_ic(int N, PDEProblem& pde, torch::Device dev) {
    auto opts = torch::TensorOptions().dtype(torch::kFloat32).device(dev);
    auto x = torch::rand({N, 1}, opts) * (pde.x_max - pde.x_min) + pde.x_min;
    auto t = torch::zeros({N, 1}, opts);
    return {x, t};
}

torch::Tensor
sample_bc(int N, PDEProblem& pde, torch::Device dev) {
    auto opts = torch::TensorOptions().dtype(torch::kFloat32).device(dev);
    return torch::rand({N, 1}, opts) * pde.t_max;
}
